picamera2/picamera2.py

```diff
@@ -1141,14 +1142,18 @@ def process_requests(self) -> None:
                     self.functions = []
                     self._future.set_exception(e)
 
-            # Some applications may want to do something to the image after they've had a change
-            # to copy it, but before it goes to the video encoder.
-            if display_request and self.post_callback:
-                self.post_callback(display_request)
+            for req in requests:
+                # Some applications may want to do something to the image after they've had a change
+                # to copy it, but before it goes to the video encoder.
+                if req not in self.completed_requests:
```

Comment:
I was curious about this, can it happen?

Reply:
Yeah, this seemed to happen with one of the opencv examples, it can take off one of the requests, via capture_buffer if I recall correctly.

Reply:
My feeling is that even if someone "eats" the capture request, we should probably still pass it to the encoder. Otherwise if someone were to capture a still while a video is recording, that frame would disappear from the video. What do you think?

Reply:
Yeah I did wonder that, however, if you remove the line
```
if req not in self.completed_requests:
```

When running -
python3.9 examples/opencv_face_detect_2.py 

We get
```
Traceback (most recent call last):
  File "/home/pi/Repositories/picamera2/picamera2/previews/q_gl_picamera2.py", line 362, in handle_requests
    request = self.picamera2.process_requests()
  File "/home/pi/Repositories/picamera2/picamera2/picamera2.py", line 1155, in process_requests
    self.post_callback(req)
  File "/home/pi/Repositories/picamera2/examples/opencv_face_detect_2.py", line 17, in draw_faces
    with MappedArray(request, "main") as m:
  File "/home/pi/Repositories/picamera2/picamera2/request.py", line 45, in __init__
    self.__buffer = _MappedBuffer(request, stream)
  File "/home/pi/Repositories/picamera2/picamera2/request.py", line 17, in __init__
    self.__fb = request.request.buffers[stream]
AttributeError: 'NoneType' object has no attribute 'buffers'
```

Reply:
I think we're probably not handling our reference counts properly for this new behaviour. The CompletedRequests get created with a reference count of 1, which gets "transferred" to "being in the completed_requests list". But if a capture comes along and removes it, then it may get recycled before we run the post_callback. So we need to tweak that a bit. Maybe that initial count of 1 gets transferred to "being in the requests lists", and being in the completed_requests list counts as an extra one? Something like that...

Reply:
Yeah I just tried bumping the refcount prior to adding to the completed requests, then decrementing after running the loop, which meant I could remove that if statement.

Just pushed changes again to this pull req.